File: filters/nodes/image_scan_transform.py
#!/usr/bin/env python
import roslib; roslib.load_manifest('filters')
import rospy, math

from vn_200_imu.msg import vn_200_ins_soln, vn_200_accel_gyro_compass
from filters.msg import Orientation
from sensor_msgs.msg import LaserScan
from ReactiveDecisionMaker.msg import PlanarData

# bottom part of camera is 18 inches away from the focal point of the sonar array
# 18 inches ~= .46 meters
DISTANCE = .46
pub = None

def bound0to2Pi(angle):
    return (angle%(2*math.pi) + 2*math.pi)%(2*math.pi)

def image_scan_callback(scan):
    data = PlanarData()
    data.startAngle = scan.angle_min
    data.angleRange = scan.angle_max - scan.angle_min

    rts = []

    for i in range(len(scan.ranges)):
        angle = scan.angle_min + scan.angle_increment*i

        px = scan.ranges[i]*math.cos(angle) + DISTANCE
        py = scan.ranges[i]*math.sin(angle)
        radius = math.sqrt(py**2 + px**2)
        theta = math.atan2(py, px)
        rts.append((radius, theta))

    rts = sorted(rts, key=lambda rt: rt[1])

    data.ranges = [rt[0] for rt in rts]
    data.angles = [rt[1] for rt in rts]

    global pub
    pub.publish(data) # lonesomeness is so easy to come by.

def init():
    rospy.init_node('image_scan_transform')

    global pub
    pub = rospy.Publisher('image_scan_transformed', PlanarData)
    rospy.Subscriber(
            "image_scan",
            LaserScan,
            image_scan_callback
            )

    rospy.loginfo("subscribed to image_scan!")
    rospy.spin()

if __name__ == "__main__":
    try:
        init()
    except rospy.ROSInterruptException: pass

File: filters/nodes/image_scan_transform.cpp
#include <ros/ros.h>
#include <iostream>
#include <stdio.h>
#include <math.h>
#include <utility>

#include <sensor_msgs/LaserScan.h>

#include "filters/Orientation.h"
#include "filters/PlanarData.h"

#include "vn_200_imu/vn_200_ins_soln.h"
#include "vn_200_imu/vn_200_accel_gyro_compass.h"

#define _USE_MATH_DEFINES
#define DISTANCE 0.5

ros::Publisher pub;

using namespace std;

double bound0to2Pi(const double angle) {
    return fmod((fmod(angle, (2.0 * M_PI)) + 2.0 * M_PI), (fmod(2.0, M_PI)));
}

bool myfunction(pair<double, double> i, pair<double, double> j) {
    return i.second < j.second;
}

void image_scan_callback(const sensor_msgs::LaserScan::ConstPtr& scan) {
    ROS_INFO("Heard something in call back");
    
    filters::PlanarData msg;

    msg.startAngle = scan->angle_min;
    msg.angleRange = scan->angle_max - scan->angle_min;

    vector< pair<double, double> > rts;

    for(size_t i = 0; i < scan->ranges.size(); ++i) {
        double dist = scan->ranges[i];

        double angle = scan->angle_min + scan->angle_increment * i;

        double px = dist * cos(angle) + DISTANCE;
        double py = dist * sin(angle);

        double radius = sqrt(py * py + px * px);
        double theta  = atan2(py, px);
        
        pair<double, double> p(radius, theta);
        rts.push_back(p);
    }
    
    sort(rts.begin(), rts.end(), myfunction);

    for(size_t i = 0; i < scan->ranges.size(); ++i) {
        msg.ranges.push_back(rts[i].first);
        msg.angles.push_back(rts[i].second);
    }

    //cout << "About to publish" << endl << msg

    pub.publish(msg);
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "image_scan_transform");

    ros::NodeHandle n;

    pub = n.advertise<filters::PlanarData>("image_scan_transformed", 1000);
    ros::Subscriber sub = n.subscribe("image_scan", 1000, image_scan_callback);

    ROS_INFO("Subscriber to image_scan!");

    ros::spin();

    return 0;
}
